jth-tests/src/test/java/com/redhat/jenkins/nodesharing/SharedNodeCloudTest.java

```diff
@@ -538,5 +543,40 @@ public void testNodeWipeoutSkip() throws Exception {
         assertThat(l, notLogged(Level.INFO, computer.getName() + ": wipeout activated"));
         // TODO Check if the workspace isn't deleted
     }
-}
 
+    @Test
+    @WithPlugin({"bouncycastle-api-2.17.hpi", "command-launcher-1.2.hpi", "durable-task-1.26.hpi",
+            "jdk-tool-1.1.hpi", "resource-disposer-0.12.hpi", "scm-api-2.2.8.hpi",
+            "structs-1.17.hpi", "workflow-api-2.30.hpi", "workflow-durable-task-step-2.22.hpi",
+            "workflow-step-api-2.16.hpi", "workflow-support-2.21.hpi", "ws-cleanup-0.35.hpi"})
```

Comment:
Why do we need all these any why they are not brought in via test-scoped dependency? Putting dozen of hpis in git ss something I certainly want to avoid.

Reply:
I've spent couple of hours to recognize that it is a totally different approach between `@WithPlugin` and with maven dependency.
The second doesn't *enable* the plugin in Jenkins nor load it, initialize it etc. - PluginManager doesn't know about it etc. The second only put it on ClassPath and that's it.
On the other hand `@WithPlugin` makes the hard work as I expect, but it is a quite limited by now (impl.), doesn't recognize maven repositories etc., only the 'resource' location is taken into the account (NPE otherwise) - see JENKINS-29181.
The problem here is the list of dependencies of ws-cleanup (plus transient dependencies which aren't available by default) is quite huge..

Reply:
Update: Probably annotate the test with `@WithPlugin` only for ws-cleanup and put the rest as maven dependencies only can work, but can we be sure it works as expected? (mainly due to not correct handling of dependecies by Jenkins core in that case)

Reply:
Ugh, I've tested my idea above, but as I stated previously, putting plugins on `classpath` only isn't enough, Jenkins doesn't recognize them, doesn't initialize them, `PluginManager` doesn't know about them etc. so dependency checking still reports many issues (e.g. java.io.IOException: Dependency workflow-api (2.25), workflow-step-api (2.13), workflow-support (2.20), durable-task (1.26), scm-api (2.2.6), structs (1.10) doesn't exist) thus we can't go with that work-around.

Reply:
You are right JTH does not install those dependencies via plugin manager (so `Jenkins.getActiveInstance().getPlugin("ws-cleanup")`) will always report it is uninstalled. And in order to install `ws-cleanup.hpi` we need to properly install all its deps this way. But it _does_ correctly initialize it in all other (important) regards.

I believe we can avoid all this we do not ask plugin manager to tell us whether the plugin is installed, but we ask instead whether the class we intend to use is loaded (`DisableDeferredWipeoutNodeProperty`). Additional advantage here is it will recognize the `ws-cleanup` is too old. See: https://github.com/jenkinsci/node-sharing-plugin/blob/7d4a4f909444d242de442ec3a0c857dda8a6d4c6/nodesharing-lib/src/main/java/com/redhat/jenkins/nodesharing/ConfigRepoAdminMonitor.java#L103-L108

Reply:
Yes, this is one way how to achieve desired behavior. However it means, we can't test anything *without* ws-cleanup at all (we can't drop that part from CP). There are couple of JIRAs reported this (originally from JGlick several years ago) issue, but there isn't a solution yet - as far as I know/try to find it. Therefore I've modified also the code SharedNodeCloud.java - L258 which works as I expected and fine (CNF not raised without installed ws-cleanup - tested; note: wondering CNF exception can't be catch...:-D.. thus `Throwable`). The cons there is we have to include necessary HPIs to jht-test as I did.

Reply:
Well, yeah. It complicates writing the test without the plugin present but I doubt that warrants adding dozen of binary files into git repository. How about using the approach I suggested and extracting the expected class name into a static property so such test can override it to something that is not a valid class (effectively simulating ws-cleanup is not installed or too old)?

Reply:
Yes, it's a good trade-off.